### ISSUE TITLE
NVBug 3262468: Bogus data produced in CUDA 11(.1) by thrust::sort_by_key with custom comparison operator over two (zipped) key arrays

### ISSUE THREAD

**Issue by Oblomov** [NVIDIA/thrust#1341]
Hello,

while testing [GPUSPH](/GPUSPH/gpusph) on CUDA 11.1 I discovered that the `thrust::sort_by_key` procedure that we use to sort the particle indices is now producing bogus data. The sort uses a custom comparator that fetches data from two different arrays (using zip_iterator etc). The sort results in one of the (sorted) key arrays being clobbered with invalid data.

I am currently in the process of finding a minimal test case, but in the mean time the bug can be observed in action on the `cuda11-thrust-sort-bug` of GPUSPH, by running

    make DamBreak3D && ./DamBreak3D --maxiter 1 | grep '64656 255 '

from the git working directly. This should produce no results (as it does on CUDA 10) if the sort is correct, but it results in numerous hits (all clobbered entries with bogus values) in CUDA 11.

Thanks for looking into this,

**Comment by allisonvacanti** [comment]
That's troubling...let me know when you've got a pure Thrust reproducer and I'll take a look.

**Comment by Oblomov** [comment]
I'm still working on trying to get a minimal reproducible, which is turning out to be harder than expected —quite surprising, given that the sorting kernel is only the second kernel to be invoked, preceded by a trivial initialization kernel.

Is there an easy way to print each CUDA API call that is being made by a program, in the order they are being made?

**Comment by Oblomov** [comment]
Interesting that by updating the NVIDIA driver now the clobbered entries have a different value (64672 instead of 64656 or 0xFCA0 instead of 0xFC90).

**Comment by Oblomov** [comment]
OK I've made some significant progress in producing the minimal test case AND found a workaround from within GPUSPH while the bug gets fixed: apparently the issue is somehow connected to the comparator functor being defined inside the scope of a template class. Moving it to the top-level scope (it's actually independent from the class, so this can and probably should be done anyway) seems to fix the issue. So there is something in the class template definition that somehow interferes with the comparator functor and its usage in thrust.

**Comment by Oblomov** [comment]
I've pushed a minimal reproducible [on the thrust-cuda11-bug branch of my titanxstall project](https://github.com/Oblomov/titanxstall/tree/thrust-cuda11-bug) (sorry for recycling an older repo —it's still relevant since that was for another issue we had with thrust ;-)).

    git clone https://github.com/Oblomov/titanxstall -b thrust-cuda11-bug
    cd titanxstall
    make test

On CUDA 10, All OK!, but with CUDA 11.1 you'll get something like:

    AFTER: 1: 64880 255 64880 255 0 22477
    terminate called after throwing an instance of 'std::runtime_error'
      what():  clobbered info

It's possible to change CUDA version by setting `NVCC=/path/to/nvcc` in `Makefile.local` (create the file if not present).

The issue can be solved simply by pulling the `ptype_has_compare` functor definition out of the `CUDAEngine` class template. Note also that if the class is not a class template the problem doesn't seem to arise, and if it's a class template that depends on some integer type (e.g. an enum) it doesn't arise either.

Hopefully this is enough information to find the issue!

**Comment by allisonvacanti** [comment]
Thanks, I'll try to take a look at this in the next few weeks.

**Comment by allisonvacanti** [comment]
I found some time to look at this, but got different results (CUDA 11.2 and g++ 7.5.0):

```
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  merge_sort: failed to synchronize: cudaErrorIllegalAddress: an illegal memory access was encountered
Aborted 
```

The source file refers to a `data.idx` file that doesn't seem to exist in the repo, that may be why.

**Comment by Oblomov** [comment]
Hello, I've added the data file to the repository. Hopefully this helps.

**Comment by allisonvacanti** [comment]
Thanks, I can repro the issue now. I'll try to debug this a bit more soon.

**Comment by Oblomov** [comment]
Thank you very much.

**Comment by allisonvacanti** [comment]
Finally got some time to look at this. It doesn't look like a Thrust bug:

Compiled the repro with nvcc 11.1.74 + gcc 7.5.0 + thrust/cub trunk, and the data is fine.
Compiled the repro with nvcc 11.2.56 + gcc 7.5.0 + thrust/cub trunk, and the data is corrupted.

I'll pass this along to the compiler folks (internal NVBug 3262468) and update this thread when we know more.

**Comment by Oblomov** [comment]
Thanks for looking into this. I does make sense that it's a compiler bug: it's probably something related to namepsaces, given that a possible workaround is to move the comparator outside of the namespace.

**Comment by allisonvacanti** [comment]
The compiler team identified the bug and a fix is planned for 11.3.

Closing -- please reopen if you see this again in 11.3 or later.

**Comment by Oblomov** [comment]
Thank you very much.

**Comment by soCzech** [comment]
@allisonvacanti Can you look at this https://github.com/pytorch/pytorch/issues/52663 ?

I thought the issue would by related to pytorch but it seems it is a thrust issue since we encountered the issue for the second time in completely different code.

I would be satisfied with your answer it will be fixed in cuda 11.3 but based on your response it seems it worked in 11.1 but in our case it does not work already in 11.0.

Thank you!

**Comment by allisonvacanti** [comment]
I don't have pytorch available to test, but there's an easy way to find out: the bug that was fixed in this issue occurred only when PTX was JITed to the target device code, but not when directly compiling device code for the target device arch with nvcc.

For example, if testing on a Volta device, try compiling with `nvcc -arch=compute_70 -code=sm_70 ...`. If the issue goes away when compiling directly for the target arch, it should be the same issue. If not, this may be a new one.

**Comment by soCzech** [comment]
@allisonvacanti We've tested it with the direct compilation and no change. After further investigation, we found in logs that we were getting simillar errors (see bellow) already back in October 2020 with Ubuntu 18.04 and Cuda 10.2 in FLANN (in `thrust::for_each` of cuda kdtree). This seems big :) I would think the issue is on our side but in case of FLANN - it was a third party python package therefore it is unlikely the issue is in our build toolchain.

```
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  for_each: failed to synchronize: cudaErrorInvalidValue: invalid argument
// or
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  for_each: failed to synchronize: cudaErrorIllegalAddress: an illegal memory access was encountered
Process finished with exit code 134 (interrupted by signal 6: SIGABRT)
```

Besides the https://github.com/pytorch/pytorch/issues/52663 issue I was able to find that deep down in our code following absurdity: `thrust::remove_if` call on (valid!) int device vector with predicate `x == -1` left some `-1` values in valid positions of the vector after the removal.
Also values `0` and `1` that were NOT present in the original vector could be found after the call in valid positions of the vector.

Could I give you Docker image and complete build instruction to replicate the https://github.com/pytorch/pytorch/issues/52663 issue so that you can investigate what could be wrong? Thank you!

**Comment by soCzech** [comment]
@allisonvacanti To reproduce one of the errors, create Dockerfile, CMakeLists.txt and main.cu files in a single directory.

Then use these build commands:
```
docker build -t test-docker-image .
docker run -it --gpus 1 -v $(pwd):/xxx -w /xxx test-docker-image bash
mkdir build
cd build
cmake ..
make
./test
```

You should get this output:
```
6 cuda:0
6 cuda:0
3 cuda:0
 15.1000
 16.2000
 17.3000
[ CUDAFloatType{3} ]
Thrust---
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  radix_sort: failed on 1st step: cudaErrorInvalidDevice: invalid device ordinal
Aborted (core dumped)
```

Interestingly, no error is raised when the line `torch::Tensor mask = (t < 20);` is replaced by `torch::Tensor mask = (t < 20).to(torch::kCPU);`. But based on our recent experiance it seems it is just pure luck and the cause is something deeper.
Output:
```
6 cuda:0
6 cpu
3 cuda:0
 15.1000
 16.2000
 17.3000
[ CUDAFloatType{3} ]
Thrust---
1 2 4 5 7 8 
a c b e f d 
END
```
---

Dockerfile:
```
FROM nvidia/cuda:11.0-devel-ubuntu20.04

ENV CUDNN_VERSION 8.0.5.39
RUN apt-get update && apt-get install -y --no-install-recommends \
    libcudnn8=$CUDNN_VERSION-1+cuda11.0 \
    libcudnn8-dev=$CUDNN_VERSION-1+cuda11.0 \
    wget \
    unzip \
    build-essential \
    libtool \
    libblas-dev \
 && apt-mark hold libcudnn8 \
 && rm -rf /var/lib/apt/lists/*

RUN wget -qO- "https://cmake.org/files/v3.17/cmake-3.17.5-Linux-x86_64.tar.gz" | tar --strip-components=1 -xz -C /usr/local

RUN wget https://download.pytorch.org/libtorch/cu110/libtorch-cxx11-abi-shared-with-deps-1.7.1%2Bcu110.zip \
        -q -O /tmp/libtorch-cxx11-abi-shared.zip \
 && unzip /tmp/libtorch-cxx11-abi-shared.zip -d /usr/local \
 && rm /tmp/libtorch-cxx11-abi-shared.zip
```

CMakeLists.txt
```
cmake_minimum_required(VERSION 3.17)
project(test LANGUAGES CUDA CXX)

set(CMAKE_PREFIX_PATH /usr/local;/usr/local/libtorch)
find_package(Torch REQUIRED)

add_executable(test main.cu)
target_compile_features(test PUBLIC cuda_std_14)
target_link_libraries(test PUBLIC ${TORCH_LIBRARIES})
```

main.cu
```
#include <thrust/execution_policy.h>
#include <torch/torch.h>
#include <iostream>
#include <thrust/host_vector.h>
#include <thrust/device_vector.h>
#include <thrust/sort.h>

int main() {
    {
        std::vector<float> numbers = {15.1, 74.9, 16.2, 75.8, 17.3, 76.7};
        torch::Tensor t = torch::from_blob(numbers.data(), {static_cast<long>(numbers.size())},
                                           torch::TensorOptions().dtype(torch::kFloat32)).to(torch::kCUDA);
        torch::Tensor mask = (t < 20);
        std::cout << t.size(0) << " " << t.device() << "\n";
        std::cout << mask.size(0) << " " << mask.device() << "\n";
        auto gpu_result = t.masked_select(mask).view({-1});
        std::cout << gpu_result.size(0) << " " << gpu_result.device() << "\n";
        std::cout << gpu_result << "\n";
    }

    std::cout << "Thrust---\n";
    thrust::host_vector<int> keysOnHost(6);
    keysOnHost[0] = 1;keysOnHost[1] = 4;keysOnHost[2] = 2;keysOnHost[3] = 8;keysOnHost[4] = 5;keysOnHost[5] = 7;

    thrust::host_vector<char> valuesOnHost(6);
    valuesOnHost[0] = 'a';valuesOnHost[1] = 'b';valuesOnHost[2] = 'c';valuesOnHost[3] = 'd';valuesOnHost[4] = 'e';valuesOnHost[5] = 'f';

    thrust::device_vector<int> keys = keysOnHost;
    thrust::device_vector<char> values = valuesOnHost;
    thrust::sort_by_key(keys.begin(), keys.end(), values.begin());
    thrust::copy(keys.begin(), keys.end(), keysOnHost.begin());
    thrust::copy(values.begin(), values.end(), valuesOnHost.begin());
    for (int i = 0; i < 6; ++i) {
        std::cout << keysOnHost[i] << " ";
    }
    std::cout << "\n";
    for (int i = 0; i < 6; ++i) {
        std::cout << valuesOnHost[i] << " ";
    }
    std::cout << "\n";
    std::cout << "END\n";
}
```

Thank you!

**Comment by allisonvacanti** [comment]
Since it's still failing with the arch flags present, this is a separate issue. Can you file a new issue with a thrust-only repro for this? It's much easier to track these down without other dependencies present.

Thanks!
